jme3-core/src/main/java/com/jme3/renderer/opengl/GLRenderer.java

```diff
@@ -1853,23 +1908,39 @@ public void setReadDrawBuffers(FrameBuffer fb) {
         if (fb == null) {
             // Set Read/Draw buffers to initial value.
             if (context.boundDrawBuf != INITIAL) {
-                gl2.glDrawBuffer(context.initialDrawBuf);
+                if(gl2!=null) {
+                    gl2.glDrawBuffer(context.initialDrawBuf);
+                } else {
+                    gles30.glDrawBuffer(context.initialDrawBuf);
+                }
```

Comment:
Would it be a bad idea to remove duplicate code like this by making GLES30 extend GL2 ? It's mostly a duplicate of that interface anyway, minus two methods... that can be overriden in the implementation to return unsupported operation exception.

Reply:
Yes, nowadays it's duplicate of GL2 plus some minor modifications although my idea is, for the long term, add everything from GLES30 and create more interfaces (GLES31, GLES32 and later when new versions are created) according to jme3 needs. 

Also there's an issue here which is the reason why I did it this way. There's plenty of code checking gl being an instance of GL2, and specially the loadCapabilities method at GLRenderer class:

`
....
        this.gl2 = gl instanceof GL2 ? (GL2)gl : null;
....
    private void loadCapabilities() {
        if (gl2 != null) {
            loadCapabilitiesGL2();
        } else {
            loadCapabilitiesES();
        }
        loadCapabilitiesCommon();
    }
`

So using GL2 interface would use GL2 caps not ES caps. Maybe we could look into a better way of checking if it's desktop gl2 or gles3 instead but may require lots of changes

Reply:
> Yes, nowadays it's duplicate of GL2 plus some minor modifications although my idea is, for the long term, add everything from GLES30 and create more interfaces (GLES31, GLES32 and later when new versions are created) according to jme3 needs.

I don't mean to remove GLES30, but to make it extend GL2. GLES30 will add the missing features and GLES31 will extend GLES30

> Also there's an issue here which is the reason why I did it this way. There's plenty of code checking gl being an instance of GL2, and specially the loadCapabilities method at GLRenderer class:

Yes, this can be an issue, but using duplicate code every time a gl2 api is called is also pretty bad. :thinking:

Reply:
So, if I understood you correctly. GLES30 extends GL2 so gl2 == | != null checks are enough for newer GL functions and the duplicated code could be removed safely. 

In loadCapabilities, we would check (gl2 != null && ! gl2 instanceof GLES30) -> loadGL2 else -> loadES, safely checking desktop vs mobile

For gles missing functions (glAlphaFunc, glPointSize and glPolygonMode) we could throw an exception and in the glrenderer check gl instance of GLES30 or just leave empty functions. Those functions are only used in GLDebugDesktop and glPolygonMode in GLRenderer once

Reply:
yes

Reply:
> Although this may be clearer and avoid duplicated code, is not strict regarding gl releases. And for the long time could generate other issues. For example, a function that's being used in jme that is defined in the specs of opengl 4.1 and in opengl es 3.2 GLES32 should extend GLES31 (clearer solution but having duplicated codelike in here) or GL4 (avoiding duplicated code but weird stuff because it's just one function defined in gles32)?

We will deal with future problems in the future. But right now now we have several of this

```
if(){
   gl.function(a,b,c)
}else{
   gles.sameFunction(a,b,c);
}
```

that are pretty bad to maintain. 
But I'm open to other solutions that can fix that.

Reply:
Maybe a cleaner solution would be making AndroidGL implement GL2 and GLES30. GLES30 would just have aditional gl functions specific for gles30 instead of extending GL2. This would avoid duplicated code, allow checking instanceof GLES30 and avoid the future problems :thinking:

Reply:
I've changed the code to match my last comment which solves the duplicated code